File: TripKitAndroid/src/main/java/com/skedgo/tripkit/BookingAction.java
package com.skedgo.tripkit;

import android.content.Intent;

import com.skedgo.tripkit.bookingproviders.BookingProvider;

import org.immutables.value.Value;

import static org.immutables.value.Value.Style.BuilderVisibility.PACKAGE;
import static org.immutables.value.Value.Style.ImplementationVisibility.PRIVATE;

@Value.Immutable
@Value.Style(visibility = PRIVATE, builderVisibility = PACKAGE)
public abstract class BookingAction {
    public static Builder builder() {
        return new BookingActionBuilder();
    }

    @BookingProvider
    public abstract int bookingProvider();

    public abstract boolean hasApp();

    public abstract Intent data();

    public interface Builder {
        Builder bookingProvider(@BookingProvider int bookingProvider);

        Builder hasApp(boolean hasApp);

        Builder data(Intent data);

        BookingAction build();
    }
}

File: TripKitAndroid/src/main/java/com/skedgo/tripkit/BookingAction.kt
package com.skedgo.tripkit

import android.content.Intent
import com.skedgo.tripkit.bookingproviders.BookingProvider
import org.immutables.value.Value.Immutable
import org.immutables.value.Value.Style
import org.immutables.value.Value.Style.BuilderVisibility.PACKAGE
import org.immutables.value.Value.Style.ImplementationVisibility.PRIVATE

@Immutable
@Style(visibility = PRIVATE, builderVisibility = PACKAGE)
abstract class BookingAction {
    @BookingProvider
    abstract fun bookingProvider(): Int

    abstract fun hasApp(): Boolean

    abstract fun data(): Intent

    interface Builder {
        fun bookingProvider(@BookingProvider bookingProvider: Int): Builder

        fun hasApp(hasApp: Boolean): Builder

        fun data(data: Intent?): Builder

        fun build(): BookingAction
    }

    companion object {
        fun builder(): Builder {
            return BookingActionBuilder()
        }
    }
}